### PR TITLE
Fix ambiguous BindAsync error

### LINKED ISSUE #36957: Fallback logic for TryParse and BindAsync is inconsistent
Minimal route handlers do not have consistent behavior between `TryParse` and `BindAsync` if the most-preferred overload has the wrong return type but there is a valid less-preferred overload.

`TryParse` will fall back to the less-preferred overload and will not throw or log that the more-preferred overload (taking an `IFormatProvider`) has the wrong return type.

`BindAsync` on the other hand will fail to find the less-preferred overload and then it will throw an `System.Reflection.AmbiguousMatchException` when trying to produce an `InvalidOperationException` for `BindAsync` being in the incorrect format.

```C#
using System.Reflection;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

// https://{host}/TryParse/test responds with "2-argument: test"
app.MapGet("/TryParse/{myTryParse}", (MyTryParse myTryParse) => myTryParse.Value);

// Uncomment the following to get System.Reflection.AmbiguousMatchException: Ambiguous match found.
//app.MapGet("/BindAsync", (MyBindAsync myBindAsync) => myBindAsync);

app.Run();

record MyTryParse(string Value)
{
    // The 3-argument overload would normally be preferred if it correctly returned bool.
    public static void TryParse(string? value, IFormatProvider formatProvider, out MyTryParse myTryParse) =>
        throw new NotImplementedException();

    // The 2-argument overload is used instead because it has the correct return type.
    public static bool TryParse(string? value, out MyTryParse? myTryParse)
    {
        myTryParse = new MyTryParse($"2-argument: {value}");
        return true;
    }
}

record MyBindAsync(int Value)
{
    // The 2-argument overload would normally be preferred if it correctly returned ValueTask<MyBindAsync>.
    public static MyBindAsync BindAsync(HttpContext context, ParameterInfo parameter) =>
        throw new NotImplementedException();

    // The 1-argument overload isn't used either even though it's valid because the 2-argument has the wrong return type.
    public static ValueTask<MyBindAsync> BindAsync(HttpContext context) =>
        throw new NotImplementedException();
}
```

And here's the full output for when `app.MapGet("/BindAsync", (MyBindAsync myBindAsync) => myBindAsync);` is uncommented.

```
Unhandled exception. System.Reflection.AmbiguousMatchException: Ambiguous match found.
   at System.RuntimeType.GetMethodImplCommon(String name, Int32 genericParameterCount, BindingFlags bindingAttr, Binder binder, CallingConventions callConv, Type[] types, ParameterModifier[] modifiers)
   at System.Type.GetMethod(String name, BindingFlags bindingAttr)
   at Microsoft.AspNetCore.Http.ParameterBindingMethodCache.<FindBindAsyncMethod>g__Finder|12_0(Type nonNullableParameterType)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.AspNetCore.Http.ParameterBindingMethodCache.FindBindAsyncMethod(ParameterInfo parameter)
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.CreateArgument(ParameterInfo parameter, FactoryContext factoryContext)
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.CreateArguments(ParameterInfo[] parameters, FactoryContext factoryContext)
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.CreateTargetableRequestDelegate(MethodInfo methodInfo, Expression targetExpression, FactoryContext factoryContext)
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.Create(Delegate handler, RequestDelegateFactoryOptions options)
   at Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.Map(IEndpointRouteBuilder endpoints, RoutePattern pattern, Delegate handler, Boolean disableInferBodyFromParameters)
   at Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapMethods(IEndpointRouteBuilder endpoints, String pattern, IEnumerable`1 httpMethods, Delegate handler)
   at Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapGet(IEndpointRouteBuilder endpoints, String pattern, Delegate handler)
   at Program.<Main>$(String[] args) in C:\dev\halter73\MinimalPlayground\Program.cs:line 10
```

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/36957

### CODE REVIEW

File: src/Shared/ParameterBindingMethodCache.cs [hunk 1]
```diff
@@ -283,7 +293,9 @@ public bool HasTryParseMethod(ParameterInfo parameter)
 
         private static MethodInfo? GetAnyMethodFromHierarchy(Type type, string name)
         {
-            var methodInfo = type.GetMethod(name, BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance | BindingFlags.FlattenHierarchy);
+            // Find first incorrectly formatted method
+            var methodInfo = type.GetMethods(BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance | BindingFlags.FlattenHierarchy)
+                .Where(methodInfo => methodInfo.Name == name).FirstOrDefault();
```
Comment (captainsafia) [inline src/Shared/ParameterBindingMethodCache.cs:298]:
```suggestion
                .FirstOrDefault(methodInfo => methodInfo.Name == name);
```

File: src/Shared/ParameterBindingMethodCache.cs [hunk 2]
```diff
@@ -168,19 +174,17 @@ public bool HasTryParseMethod(ParameterInfo parameter)
             {
                 var hasParameterInfo = true;
                 // There should only be one BindAsync method with these parameters since C# does not allow overloading on return type.
-                var methodInfo = GetStaticMethodFromHierarchy(nonNullableParameterType, "BindAsync", new[] { typeof(HttpContext), typeof(ParameterInfo) });
-                if (methodInfo is null)
+                var methodInfo = GetStaticMethodFromHierarchy(nonNullableParameterType, "BindAsync", new[] { typeof(HttpContext), typeof(ParameterInfo) }, ValidateReturnType);
+                if (methodInfo is null || !ValidateReturnType(methodInfo))
```
Comment (captainsafia) [inline src/Shared/ParameterBindingMethodCache.cs:178]: So if there is no `BindAsync` method or it doesn't produce the a valid return type then we fallback to `BindAsync(HttpContext)` without a parameter info. What is this functionally doing?

**Review by captainsafia — COMMENTED** [review]
LGTM, but will defer a final review to a SME. :)

File: src/Shared/ParameterBindingMethodCache.cs (hunk 2 above)
Comment (BrennanConroy) [reply src/Shared/ParameterBindingMethodCache.cs:178]: > or it doesn't produce the a valid return type

I thought I removed that `||` statement, it's now passed directly to `GetStaticMethodFromHierarchy` so isn't needed anymore.

> What is this functionally doing?

Falling back from `BindAsync(HttpContext, ParameterInfo)` to `BindAsync(HttpContext)`.

**Review by halter73 — APPROVED** [review]

**Pull request merged**